src/python/pants/engine/target.py

```diff
@@ -757,42 +722,101 @@ def create(cls: Type[_CWO], target_with_origin: TargetWithOrigin) -> _CWO:
             **_get_config_fields_from_target(cls, tgt),
         )
 
-    @final
-    @classmethod
-    def group_targets_to_valid_subclass_configs(
-        cls: Type[_CWO],
-        targets_with_origins: TargetsWithOrigins,
+
+@frozen_after_init
+@dataclass(unsafe_hash=True)
+class TargetsToValidConfigurations:
+    mapping: FrozenDict[TargetWithOrigin, Tuple[_AbstractConfiguration, ...]]
+
+    def __init__(
+        self, mapping: Mapping[TargetWithOrigin, Iterable[_AbstractConfiguration]]
+    ) -> None:
+        self.mapping = FrozenDict(
+            {tgt_with_origin: tuple(configs) for tgt_with_origin, configs in mapping.items()}
+        )
+
+    @memoized_property
```

Comment:
I think it makes sense to memoize this? The idea is to avoid having to re-iterate over the collection multiple times.